enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+
+### Workflow Description
+
+End users have 2 possible variations for the creation of the route:
+
+- Create a route for the user workload and this is completely managed by the end user.
+- Create an ingress for the user workload and a managed route is created automatically
+  by the ingress-to-route controller.
+
+Both these workflows will support integrating with third party certificate management
+solutions like cert-manager with the secret-reference annotation described under [API Extensions](#api-extensions).
+
+**When a Route is directly used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Route is going to be created.
+- To expose a user workload, the user would create a new Route with the
+  secret-reference annotation referencing the generated secret that was created
+  in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair, the secret-injector controller copies the certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+**When an Ingress is used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Ingress is going to be created.
+- To expose a user workload, a new Ingress with the generated secret
+  referenced in `.spec.tls.secretName` needs to be created.
+- If the secret CR that is referenced exists and has a successfully generated
+  cert/key pair, the ingress-to-route controller copies this certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+_Note_: The same workflow would apply to updating an existing Route to integrate
+with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
+and `.spec.tls.key` in the route CR since the secret-reference annotation gets a
+higher preference.
+
+#### Certificate Life cycle
+
+Post integrating with third-party solutions like cert-manager, the end user can also
+disable automatic certificate management. This can be done in 2 ways,
+
+- The end user can delete the secret-reference annotation on the Route.
+- The end user deletes the secret that was associated with the Certificate CR.
+
+**Deleting the secret-reference annotation on the Route**
+The end user can delete the annotation on the Route, this will result
+in clearing `.spec.tls.certificate` and `.spec.tls.key` and resort to
+using the default certificates that are generated by the cluster-ingress-operator.
```

Comment:
I suppose we could have an annotation like `last-applied-configuration` to store the old data from `spec.tls.certificate` and `spec.tls.key`.  Would that be less confusing?

Reply:
Not sure how I feel about an annotation API, in general we steer away from them, but I guess this would be considered "internal" so it's probably ok? The idea of storing it to restore later seems beneficial though, +1 to that